### ISSUE TITLE
Reduce framework size

### ISSUE THREAD

**Issue by Banck** [grpc/grpc-swift#1329]
Hello. 
We are using grpc in project, which has App Clip,so that’s why we care about the size of the application (there is limit of 10 mb of uncompressed app). 
grpc integration (via spm) increased uncompressed app size by about 6.5 mb. 
Maybe there are any solutions to reduce framework size?
Thanks.

**Comment by Lukasa** [comment]
Can I recommend stripping the app binary? Much of the size is likely to be symbols, and these can easily be removed by stripping them. Otherwise, you can try enabling `-Osize` optimizations. Beyond that, there are no ways to reduce binary size in SwiftPM.

**Comment by Banck** [comment]
-0size doesn't help. It reduced app clip size from 9.4 mb to 9.3 mb)
Would't like to use a proxy for the grpc in App Clip, but grpc integration increases size a lot(

**Comment by Banck** [comment]
Tried switch from SPM to Cocoapods and AppClip uncompressed size increased from 9.4 mb to 11.5 mb 😨

**Comment by Lukasa** [comment]
Have you tried stripping the binary?

**Comment by Banck** [comment]
> Have you tried stripping the binary?
I have `bitcode`, `Strip linked product`, `Dead code Stripping` enabled.
I've tried enable `Strip debug symbols during copy` for main target and app clip target, but app size didn't change.

App clip size with `bitcode`, `Strip linked product`, `Dead code Stripping` enabled:
App Thinning Size Report Supported variant descriptors: Universal
1) Cocoapods with `use_modular_headers!`
App size: 3,9 MB compressed, **8,9** MB uncompressed

2)  SPM
<img width="529" alt="image" src="https://user-images.githubusercontent.com/6259254/148892369-dfdded4a-08e0-4f82-ad0d-d2eb2a60812f.png">

 a) [-O]
App size: 4,3 MB compressed, **9,9** MB uncompressed

 b) Swift compiler: [-Osize], Apple clang Optimization level: [-0s]
App size: 4,3 MB compressed, **9,8** MB uncompressed

 c) Swift compiler: [-Osize], Apple clang Optimization level: [-0z]
 App size: 4,3 MB compressed, 9,8 MB uncompressed

 
 Don't know what can I do else to reduce grpc lib size

**Comment by Lukasa** [comment]
Thanks @Banck. Can you run the [`analyze_code_size.py` script from Swift](https://raw.githubusercontent.com/apple/swift/main/utils/analyze_code_size.py) against your app clip and provide the output here? The invocation is probably something like: `analyze_code_size.py -categorize <path/to/your/binary>`.

**Comment by Banck** [comment]
> Thanks @Banck. Can you run the [`analyze_code_size.py` script from Swift](https://raw.githubusercontent.com/apple/swift/main/utils/analyze_code_size.py) against your app clip and provide the output here? The invocation is probably something like: `analyze_code_size.py -categorize <path/to/your/binary>`.

1) Cocoapods
[Cocoapods.txt](https://github.com/grpc/grpc-swift/files/7848148/Cocoapods.txt)
2) SPM
[SPM.txt](https://github.com/grpc/grpc-swift/files/7848150/SPM.txt)

**Comment by Lukasa** [comment]
Hah, sorry, can you remove the symbol stripping before you do this? That data is not all that useful without the symbols, as all we can tell is that we're using the code size mostly with code.

I also want to point out apple/swift-protobuf#1204.

**Comment by Banck** [comment]
> Hah, sorry, can you remove the symbol stripping before you do this? That data is not all that useful without the symbols, as all we can tell is that we're using the code size mostly with code.
> 
> I also want to point out [apple/swift-protobuf#1204](https://github.com/apple/swift-protobuf/issues/1204).


Disabled Bitcode and applied these settings:
<img width="530" alt="image" src="https://user-images.githubusercontent.com/6259254/148998519-7e41c4a8-9a7c-4f92-9f36-4f8c76427393.png">

1) Cocoapods:
App size: 6,4 MB compressed, 24,1 MB uncompressed
[Cocoapods.txt](https://github.com/grpc/grpc-swift/files/7848977/Cocoapods.txt)

2) SPM
<img width="488" alt="image" src="https://user-images.githubusercontent.com/6259254/148999584-44c74001-98ce-4e7a-9cdb-aecb3a80ded4.png">

App size: 7 MB compressed, 26,3 MB uncompressed
[SPM.txt](https://github.com/grpc/grpc-swift/files/7849018/SPM.txt)


let me know if I did something wrong or if something needs to be done additionally. Thank you

**Comment by Banck** [comment]
> I also want to point out [apple/swift-protobuf#1204](https://github.com/apple/swift-protobuf/issues/1204).

When I wrote the first message, I measured app size, then integrated grpc with 1 simple proto file, and measured again. The size difference was around 6.5 mb.

**Comment by Lukasa** [comment]
Is it possible for you to compare with only the impact from swift-protobuf? I'd like to isolate how much is grpc-swift. Keep your `.proto` files the same, but don't generate the grpc sources, only the regular protobuf ones?

**Comment by Banck** [comment]
> Is it possible for you to compare with only the impact from swift-protobuf? I'd like to isolate how much is grpc-swift. Keep your `.proto` files the same, but don't generate the grpc sources, only the regular protobuf ones?

I measured app size with bitcode and stripping and integrated grpc via SPM:
App size: 4,3 MB compressed, 9,9 MB uncompressed

Then removed all services, generated before and generated only models via comand:
`protoc *.proto --swift_out=../Common/Models/Network/GRPC
`
1) Commented all my code related with services. Commented import GRPC in files:
App size: 4,3 MB compressed, 9,8 MB uncompressed

2) removed GRPC from spm
 a) added NIO, NIOHTTP2, Logging and protobuf:
<img width="179" alt="image" src="https://user-images.githubusercontent.com/6259254/149008176-fe87c1b9-857b-4e08-91df-db9175b48d55.png">

 App size: 3,1 MB compressed, 7,1 MB uncompressed

 b) added these libs:
 
<img width="487" alt="image" src="https://user-images.githubusercontent.com/6259254/149006877-3b8dc51b-b232-492b-845d-0cb36ee588ab.png">

App size: 3,3 MB compressed, 7,5 MB uncompressed

**Comment by Banck** [comment]
And sure, only with SwiftProtobuf:
App size: 2,1 MB compressed, 4,3 MB uncompressed

**Comment by Lukasa** [comment]
Ok, so based on those numbers, the big code size contributor appears to be `NIO`. This is not necessarily a huge surprise to me: `NIOPosix` is a particularly large library. Can you clarify: is your project directly importing and depending on `NIO`, or on `NIOCore`?

**Comment by Banck** [comment]
Yes, I do import NIO and import NIOCore, because grpc returns EventLoopFuture and I working with it, which is part of swiftNIO.

**Comment by Banck** [comment]
> For my own future reference:
> 
> * Protobuf only: 2.1 MB compressed, 4.3 MB uncompressed
> * Protobuf, NIO, NIOHTTP2, Logging: 3.1 MB compressed, 7.1 MB uncompressed
> * All extra libraries including grpc: 3.3 MB compressed, 7.5 MB uncompressed
> 
> This sure _implies_ that NIO is the bulk of the weight. NIOCore should not be that heavy, but NIOPosix may well be.

7.5 mb - it's without grpc. 
Grpc with all dependecies(nio, protobuf): 
App size: 4,3 MB compressed, 9,8 MB uncompressed

**Comment by Lukasa** [comment]
@Banck It's part of SwiftNIO, but we recently split the `NIO` module up into _three_ modules: `NIOCore`, `NIOEmbedded`, and `NIOPosix`. This is because for iOS apps, `NIOPosix` is not really necessary, but it still costs code size.

Now, unfortunately grpc itself still requires `NIOPosix`. We don't really have great tools for making that functionality conditional at the moment, so it's extremely hard to remove.

**Comment by Banck** [comment]
I've noticed, that through SPM I can install grpc without swift-nio-ssl, cause Package.swift contains:
```
let includeNIOSSL = ProcessInfo.processInfo.environment["GRPC_NO_NIO_SSL"] == nil
.package(
    url: "https://github.com/apple/swift-nio-ssl.git",
    from: "2.14.0"
  ),
  if: includeNIOSSL
```
And this allowed us to reduce the size of the application by 2mb
from 
`App size: 4,8 MB compressed, 11,3 MB uncompressed` 
to
`App size: 4 MB compressed, 9,8 MB uncompressed`

If install grpc through cocoapods app size is
`App size: 4,2 MB compressed, 10 MB uncompressed`

I just wonder how can install `grpc` through cocoapods also without swift-nio-ssl ? Or pods already stripped nio-ssl...
